8290041: ModuleDescriptor.hashCode is inconsistent

jaikiran opened the pull request:
Can I please get a review of this change which proposes to fix https://bugs.openjdk.org/browse/JDK-8290041? 

As noted by the reporter, the current implementation is buggy since the calculation can result in a different value of the hashcode depending on the order of iteration of the `Modifier`s. The commit in this PR changes that computation to produce consistent result irrespective of the order in which the `Modifier`s (enum) is iterated upon.

A new test has been added which reproduces the issue and verifies the fix.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290041](https://bugs.openjdk.org/browse/JDK-8290041): ModuleDescriptor.hashCode is inconsistent


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9790/head:pull/9790` \
`$ git checkout pull/9790`

Update a local copy of the PR: \
`$ git checkout pull/9790` \
`$ git pull https://git.openjdk.org/jdk pull/9790/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9790`

View PR using the GUI difftool: \
`$ git pr show -t 9790`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9790.diff">https://git.openjdk.org/jdk/pull/9790.diff</a>

</details>

plevart commented on src/java.base/share/classes/java/lang/module/ModuleDescriptor.java:
```diff
@@ -2627,7 +2627,7 @@ private static <M> String toString(Set<M> mods, String what) {
         int h = 0;
         for (Enum<?> e : enums) {
-            h = h * 43 + Objects.hashCode(e.name());
+            h = h + Objects.hashCode(e.name()) * 43;
```
why * 43 ? Isn't String.hashCode good enough?

AlanBateman replied:
A sum of e.name().hashCode() should be okay here.

plevart commented:
Hi,
Sorry for joining late. I noticed that modsHashCode is a private method and it takes Iterable<? extends Enum<?>> as a parameter. But everywhere it is called, a Set<? extends Enum<?>> is passed to it. So why not just calling the set.hashCode() and be done with it? It may be even more optimal in some cases.

AlanBateman commented:
> Sorry for joining late. I noticed that modsHashCode is a private method and it takes Iterable> as a parameter. But everywhere it is called, a Set> is passed to it. So why not just calling the set.hashCode() and be done with it? It may be even more optimal in some cases.

There's a lot going on here as the hash code is generated at link time and we also want it to be reproducible across builds. We also had a CDS bug with the  the way that Enum objects were archived - Ioi has fixed that in JDK 19 so we have one less thing to be concerned about now.

plevart commented:
> > Sorry for joining late. I noticed that modsHashCode is a private method and it takes Iterable as a parameter. But everywhere it is called, a Set is passed to it. So why not just calling the set.hashCode() and be done with it? It may be even more optimal in some cases.
> 
> There's a lot going on here as the hash code is generated at link time and we also want it to be reproducible across builds. We also had a CDS bug with the the way that Enum objects were archived - Ioi has fixed that in JDK 19 so we have one less thing to be concerned about now.

Ah, I see. Enum.hashCode() is identity hash code. I have been beaten by that in the past and apparently this is a sore point for many. I wonder if it could be changed. Is unstable hashCode (accross JVM instances) a feature here or just unfortunate implementation detail?

AlanBateman commented on test/jdk/java/lang/module/ModuleDescriptorHashCodeTest.java:
```diff
@@ -63,6 +62,45 @@ public void testBootModuleDescriptor() throws Exception {
+
+        // important to use Set.of() (i.e. backed by immutable set) to reproduce the issue
+        final Set<ModuleDescriptor.Opens.Modifier> mod1 = Set.of(
+                ModuleDescriptor.Opens.Modifier.SYNTHETIC, ModuleDescriptor.Opens.Modifier.MANDATED);
```
Thieseshould probably be "mods1" and "mods2" rather than "mod1" and "mod2" as they are a set of modifiers rather than a single modifier.

jaikiran replied:
Fixed in the latest version of this PR.

AlanBateman commented on test/jdk/java/lang/module/ModuleDescriptorHashCodeTest.java:
```diff
@@ -63,6 +62,45 @@ public void testBootModuleDescriptor() throws Exception {
+     * descriptors, have the same hashcode.
+     */
+    @Test
+    public void testModifiersOrdering() throws Exception {
```
Are you planning to expand the test to cover Requires.Modifier and Exports.Modifier too?

jaikiran commented:
Hello Alan, Peter, you are right - the hashcode computation and do away with the `* 43` part. I've updated this PR to do that change. I've also updated the tests to add coverage for the exports and requires modifiers as suggested by Alan.

AlanBateman commented on test/jdk/java/lang/module/ModuleDescriptorHashCodeTest.java:
```diff
@@ -73,4 +171,23 @@ private static ModuleDescriptor fromModuleInfoClass(Module module) throws IOExce
+            Set<ModuleDescriptor.Requires.Modifier> reqsModifiers,
+            Set<ModuleDescriptor.Exports.Modifier> expsModifiers) {
+
+        final var mdb = ModuleDescriptor.newModule("foobar");
```
Does "mdb" stand for something here?

jaikiran replied:
Hello Alan, that was supposed to mean module descriptor builder. I will rename it to `builder` to be clear.

AlanBateman commented on test/jdk/java/lang/module/ModuleDescriptorHashCodeTest.java:
```diff
@@ -63,6 +62,105 @@ public void testBootModuleDescriptor() throws Exception {
+        // important to use Set.of() (i.e. backed by immutable set) to reproduce the issue
+        final Set<ModuleDescriptor.Opens.Modifier> mods2 = Set.of(
+                ModuleDescriptor.Opens.Modifier.MANDATED, ModuleDescriptor.Opens.Modifier.SYNTHETIC);
+        final ModuleDescriptor desc2 = createModuleDescriptor(mods2, null, null);
```
If you static important Requires, Exports and Opens then it should mean you can fit these onto a single line and might be easier to read, e.g.
```
var mods2 = Set.of(Opens.Modifier.MANDATED, Opens.Modifier.SYNTHETIC);
```

AlanBateman commented on src/java.base/share/classes/java/lang/module/ModuleDescriptor.java:
```diff
@@ -2627,7 +2627,7 @@ private static <M> String toString(Set<M> mods, String what) {
         int h = 0;
         for (Enum<?> e : enums) {
-            h = h * 43 + Objects.hashCode(e.name());
+            h += e.name().hashCode();
```
Good

jaikiran commented:
I've now updated the PR to rename `mdb` to `builder` and add relevant imports to reduce line lengths where those enums were referenced.

AlanBateman reviewed (approved):
Thanks for the updates, I think it looks good.

jaikiran commented:
Thank you Alan and Peter for the reviews. tier1, tier2 and tier3 testing passed with these changes.

jaikiran commented:
/integrate
